app/src/test/java/com/github/se/eventradar/EventOverviewViewModelTest.kt

```diff
@@ -441,4 +443,104 @@ class EventsOverviewViewModelTest {
+
+    // Setup your ViewModel with mocked dependencies
+    // automatically triggers the 'init' block, which calls 'observeAllEvents()'
+
+    val viewModel = EventsOverviewViewModel(eventRepository, userRepository)
+
+    delay(100)
+    eventRepository.addEvent(testEvent2)
+
+    (eventRepository as MockEventRepository)
+        .eventsFlow
+        .emit(Resource.Success(listOf(testEvent1, testEvent2)))
+
+    assert(viewModel.uiState.value.eventList.allEvents.containsAll(listOf(testEvent1, testEvent2)))
+  }
+
+  @Test
+  fun testObserveAllEventsFailure() = runTest {
+    mockkStatic(Log::class)
+    every { Log.d(any(), any()) } returns 0
+
+    val exception = Exception("Network error")
+
+    (eventRepository as MockEventRepository).eventsFlow.emit(Resource.Failure(exception))
+
+    EventsOverviewViewModel(eventRepository, userRepository)
+
+    delay(500)
+
+    val expectedLogMessage = "Failed to fetch events: $exception"
+
+    verify { Log.d("EventsOverviewViewModel", expectedLogMessage) }
+
+    unmockkAll()
+  }
+
+  @Test
+  fun testObserveUpcomingEventsSuccess() = runTest {
+    val testEvent1 = mockEvent.copy(fireBaseID = "1", attendeeList = mutableListOf("user1"))
+    val testEvent2 = mockEvent.copy(fireBaseID = "2", attendeeList = mutableListOf("user1"))
+    val testEvent3 = mockEvent.copy(fireBaseID = "3", attendeeList = mutableListOf("user2"))
+    eventRepository.addEvent(testEvent1)
+    eventRepository.addEvent(testEvent2)
+    eventRepository.addEvent(testEvent3)
+
+    (userRepository as MockUserRepository).updateCurrentUserId("user1")
+
+    val viewModel = EventsOverviewViewModel(eventRepository, userRepository)
+
+    (eventRepository as MockEventRepository)
+        .eventsFlow
+        .emit(Resource.Success(listOf(testEvent1, testEvent2, testEvent3)))
+
+    assert(viewModel.uiState.value.upcomingEventList.allEvents.size == 2)
+    assert(
+        viewModel.uiState.value.upcomingEventList.allEvents.containsAll(
+            listOf(testEvent1, testEvent2)))
+  }
+
+  @Test
+  fun testObserveUpcomingEventsUserIdFetchFailure() = runTest {
+    mockkStatic(Log::class)
+    every { Log.d(any(), any()) } returns 0
+
+    val exception = Exception("No user currently signed in")
+    (userRepository as MockUserRepository).updateCurrentUserId(null)
+
+    EventsOverviewViewModel(eventRepository, userRepository)
+
+    delay(500)
+
+    verify { Log.d("EventsOverviewViewModel", "Error fetching user ID: ${exception.message}") }
+    unmockkAll()
+  }
+
+  @Test
+  fun testObserveUpcomingEventsFetchFailure() = runTest {
+    mockkStatic(Log::class)
+    every { Log.d(any(), any()) } returns 0
+
+    (userRepository as MockUserRepository).updateCurrentUserId("user1")
+    val exception = Exception("Network error")
+    (eventRepository as MockEventRepository).eventsFlow.emit(Resource.Failure(exception))
+
+    EventsOverviewViewModel(eventRepository, userRepository)
+
+    delay(500)
```

Comment:
nit: not sure exactly but these delay calls might not be good for regression (maybe at some point in the future it will take the `emit()` function 600 milliseconds to emit a failure instead of 500). Just a thought for future reference.

Reply:
It's something to keep in mind for future optimizations, as relying on fixed delays can introduce potential flakiness in tests if the underlying timings change.
 Note: It also worked with 100ms. I put 500 to be safe.